polars/polars-core/src/chunked_array/ops/sort/argsort_multiple.rs

```diff
@@ -23,24 +24,43 @@ pub(crate) fn args_validate<T: PolarsDataType>(
     Ok(())
 }
 
-pub(crate) fn argsort_multiple_impl<T: PartialOrd>(
-    mut vals: Vec<(IdxSize, Option<T>)>,
+fn sort_cmp<T: PartialOrd + IsFloat + Copy>(a: &T, b: &T) -> Ordering {
+    if T::is_float() {
+        match (a.is_nan(), b.is_nan()) {
+            // safety: we checked nans
+            (false, false) => unsafe { a.partial_cmp(b).unwrap_unchecked() },
+            (true, true) => Ordering::Equal,
+            (true, false) => Ordering::Greater,
+            (false, true) => Ordering::Less,
+        }
+    } else {
+        // no floats, so we can compare unchecked
+        unsafe { a.partial_cmp(b).unwrap_unchecked() }
```

Comment:
I suspect this means `IsFloat` should be an unsafe trait (unsafe to implement). Otherwise safe user code can implement it incorrectly on some local type, leading to a soundness issue.

Reply:
Good call. We should make that trait unsafe indeed.

Reply:
Documentation could also help understand what `IsFloat` means.

For example, I could define a local type for 2D points on the integer grid (`struct Point { x: i32, y: i32 }`). This struct could have a partial ordering implementation (`a > b` iff `a.x > b.x && a.y > b.y`), returning `None` in some cases.

Implementing `IsFloat` _sounds_ easy in that case (it's not a float), but even then we shouldn't be implementing this trait at all here, since there is no easy way to properly behave in this sort function.

Reply:
It's not meant to be public. It's only for polars physical  types.

It is public because its in one of our subcrates, but any user implementation can never be used here. Because we only get here via polars physical types.


But for correctness I will make it an unsafe trait.

Reply:
Is it actually impossible to call this method on user-defined types?

`PolarsNumericType` can be implemented on local types. I haven't tried yet, but is it really impossible to build a `ChunkedArray<LocalType>`? Or a `ChunkedArray<Option<LocalType>>`?

Reply:
Ah, yes that would indeed be an elegant solution for this!

Reply:
> PolarsNumericType can be implemented on local types. I haven't tried yet, but is it really impossible to build a ChunkedArray<LocalType> (or a ChunkedArray<Option<LocalType>>)?

Yes, that would blow up very quickly. You would not get your type into arrow arrays.